// @angular-generic-table/core/esm5/components/gt-table-info.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ChangeDetectorRef, Input } from '@angular/core';
import { GenericTableComponent } from './generic-table.component';
var GtTableInfoComponent = /** @class */ (function () {
    function GtTableInfoComponent(_changeDetectionRef) {
        this._changeDetectionRef = _changeDetectionRef;
    }
    /**
     * @return {?}
     */
    GtTableInfoComponent.prototype.ngAfterViewChecked = /**
     * @return {?}
     */
    function () {
        this._changeDetectionRef.detectChanges();
    };
    GtTableInfoComponent.decorators = [
        { type: Component, args: [{
                    selector: 'gt-table-info',
                    template: "<span *ngIf=\"genericTable.gtInfo\">{{(customText? customText:genericTable.gtTexts) | gtTableInfo:\n\t\tgenericTable.gtInfo:\n\t\tgenericTable.gtInfo.recordsAfterSearch:\n\t\tgenericTable.gtInfo.recordFrom:\n\t\tgenericTable.gtInfo.recordTo:\n\t\tgenericTable.gtInfo.recordsAll:\n\t\tgenericTable.gtTexts.loading:\n\t\tgenericTable.gtTexts.tableInfoAfterSearch}}</span>"
                },] },
    ];
    /** @nocollapse */
    GtTableInfoComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef, },
    ]; };
    GtTableInfoComponent.propDecorators = {
        "genericTable": [{ type: Input },],
        "customText": [{ type: Input },],
    };
    return GtTableInfoComponent;
}());
export { GtTableInfoComponent };
function GtTableInfoComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    GtTableInfoComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    GtTableInfoComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    GtTableInfoComponent.propDecorators;
    /** @type {?} */
    GtTableInfoComponent.prototype.genericTable;
    /** @type {?} */
    GtTableInfoComponent.prototype.customText;
    /** @type {?} */
    GtTableInfoComponent.prototype._changeDetectionRef;
}
import { Pipe } from '@angular/core';
var TableInfoPipe = /** @class */ (function () {
    function TableInfoPipe() {
    }
    /**
     * @param {?} texts
     * @param {?} keys
     * @param {?} recordsAfterSearch
     * @param {?} recordFrom
     * @param {?} recordTo
     * @param {?} recordsAll
     * @param {?} loading
     * @param {?} tableInfoAfterSearch
     * @return {?}
     */
    TableInfoPipe.prototype.transform = /**
     * @param {?} texts
     * @param {?} keys
     * @param {?} recordsAfterSearch
     * @param {?} recordFrom
     * @param {?} recordTo
     * @param {?} recordsAll
     * @param {?} loading
     * @param {?} tableInfoAfterSearch
     * @return {?}
     */
    function (texts, keys, recordsAfterSearch, recordFrom, recordTo, recordsAll, loading, tableInfoAfterSearch) {
        var /** @type {?} */ text = typeof texts === 'object' ? texts.tableInfo : '';
        if (keys.recordsAfterSearch !== keys.recordsAll) {
            text = typeof texts === 'object' ? texts.tableInfoAfterSearch : '';
        }
        for (var /** @type {?} */ key in keys) {
            if (keys.hasOwnProperty(key)) {
                var /** @type {?} */ searchString = new RegExp('#' + key, 'g');
                text = text.replace(searchString, keys[key]);
            }
        }
        return text;
    };
    TableInfoPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'gtTableInfo'
                },] },
    ];
    return TableInfoPipe;
}());
export { TableInfoPipe };
function TableInfoPipe_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    TableInfoPipe.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    TableInfoPipe.ctorParameters;
}
