// @angular-generic-table/core/esm5/components/gt-checkbox/gt-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, EventEmitter, HostBinding, Input, Output } from '@angular/core';
var GtCheckboxComponent = /** @class */ (function () {
    function GtCheckboxComponent() {
        this.class = 'd-flex justify-content-end';
        this.changed = new EventEmitter();
    }
    Object.defineProperty(GtCheckboxComponent.prototype, "initialValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._initialValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._initialValue = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GtCheckboxComponent.prototype, "checked", {
        get: /**
         * @return {?}
         */
        function () {
            return this._checked;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._checked = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    GtCheckboxComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} checked
     * @return {?}
     */
    GtCheckboxComponent.prototype.toggle = /**
     * @param {?} checked
     * @return {?}
     */
    function (checked) {
        this.changed.emit(this.checked);
    };
    GtCheckboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'gt-checkbox',
                    template: "\n    <label class=\"custom-control custom-checkbox p-0\" (click)=\"$event.stopPropagation()\">\n      <input #checkbox type=\"checkbox\" class=\"custom-control-input\" [(checked)]=\"checked\" (change)=\"toggle($event);\">\n      <span class=\"custom-control-label\"></span>\n    </label>\n  ",
                    styles: []
                },] },
    ];
    /** @nocollapse */
    GtCheckboxComponent.ctorParameters = function () { return []; };
    GtCheckboxComponent.propDecorators = {
        "class": [{ type: HostBinding, args: ['class',] },],
        "initialValue": [{ type: Input },],
        "checked": [{ type: Input },],
        "changed": [{ type: Output },],
    };
    return GtCheckboxComponent;
}());
export { GtCheckboxComponent };
function GtCheckboxComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    GtCheckboxComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    GtCheckboxComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    GtCheckboxComponent.propDecorators;
    /** @type {?} */
    GtCheckboxComponent.prototype.class;
    /** @type {?} */
    GtCheckboxComponent.prototype._checked;
    /** @type {?} */
    GtCheckboxComponent.prototype._initialValue;
    /** @type {?} */
    GtCheckboxComponent.prototype.changed;
}
