// @angular-generic-table/core/esm2015/components/gt-table-info.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ChangeDetectorRef, Input } from '@angular/core';
import { GenericTableComponent } from './generic-table.component';
export class GtTableInfoComponent {
    /**
     * @param {?} _changeDetectionRef
     */
    constructor(_changeDetectionRef) {
        this._changeDetectionRef = _changeDetectionRef;
    }
    /**
     * @return {?}
     */
    ngAfterViewChecked() {
        this._changeDetectionRef.detectChanges();
    }
}
GtTableInfoComponent.decorators = [
    { type: Component, args: [{
                selector: 'gt-table-info',
                template: `<span *ngIf="genericTable.gtInfo">{{(customText? customText:genericTable.gtTexts) | gtTableInfo:
		genericTable.gtInfo:
		genericTable.gtInfo.recordsAfterSearch:
		genericTable.gtInfo.recordFrom:
		genericTable.gtInfo.recordTo:
		genericTable.gtInfo.recordsAll:
		genericTable.gtTexts.loading:
		genericTable.gtTexts.tableInfoAfterSearch}}</span>`
            },] },
];
/** @nocollapse */
GtTableInfoComponent.ctorParameters = () => [
    { type: ChangeDetectorRef, },
];
GtTableInfoComponent.propDecorators = {
    "genericTable": [{ type: Input },],
    "customText": [{ type: Input },],
};
function GtTableInfoComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    GtTableInfoComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    GtTableInfoComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    GtTableInfoComponent.propDecorators;
    /** @type {?} */
    GtTableInfoComponent.prototype.genericTable;
    /** @type {?} */
    GtTableInfoComponent.prototype.customText;
    /** @type {?} */
    GtTableInfoComponent.prototype._changeDetectionRef;
}
import { Pipe } from '@angular/core';
export class TableInfoPipe {
    /**
     * @param {?} texts
     * @param {?} keys
     * @param {?} recordsAfterSearch
     * @param {?} recordFrom
     * @param {?} recordTo
     * @param {?} recordsAll
     * @param {?} loading
     * @param {?} tableInfoAfterSearch
     * @return {?}
     */
    transform(texts, keys, recordsAfterSearch, recordFrom, recordTo, recordsAll, loading, tableInfoAfterSearch) {
        let /** @type {?} */ text = typeof texts === 'object' ? texts.tableInfo : '';
        if (keys.recordsAfterSearch !== keys.recordsAll) {
            text = typeof texts === 'object' ? texts.tableInfoAfterSearch : '';
        }
        for (const /** @type {?} */ key in keys) {
            if (keys.hasOwnProperty(key)) {
                const /** @type {?} */ searchString = new RegExp('#' + key, 'g');
                text = text.replace(searchString, keys[key]);
            }
        }
        return text;
    }
}
TableInfoPipe.decorators = [
    { type: Pipe, args: [{
                name: 'gtTableInfo'
            },] },
];
function TableInfoPipe_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    TableInfoPipe.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    TableInfoPipe.ctorParameters;
}
